✨ Feat: 스터디 그룹 CRU API

dev-s-kr opened the pull request:
## ✅ PR 요약
- 작업 요약: 스터디 그룹 생성, 조회, 상세 조회, 수정 API 시리얼라이저, 모델, 뷰 작성

## 📄 상세 내용
- [ ] 스터디 그룹 CRU 시리얼라이저 수정
- [ ] 스터디 그룹 CRU API 작성
- [ ] 스터디 그룹 페이지네이션 기능 추가
- [ ] 스터디 그룹 조회에 쿼리 파라미터 조건 추가
-----
- [ ] 스터디 그룹 테스트 코드 작성
- [ ] 노출되는 id -> uuid로 수정
- [ ] 목록 조회에서 금액 -> 원가와 할인가 둘 다 응답
- [ ] urls.py group_id -> group_uuid로 수정
-----
- [ ] bulk_create 사용으로 쿼리문 반복 방지
- [ ] 모델에 ManyToMany로 연결, related_name 변경
- [ ] id 받아오는 부분 uuid로 수정
- [ ] ManyToMany의 set 메서드 사용하여 코드 간결화
- [ ] 사용하지 않는 get_lectures 코드 삭제
- [ ] transaction을 사용하여 데이터 무결성 보장

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

dev-s-kr committed:
✨ Feat: 스터디 그룹 CRU API

스터디 그룹 생성, 조회, 상세 조회, 수정 API 시리얼라이저, 모델, 뷰 작성

dev-s-kr committed:
✨ Feat: 스터디 그룹 CRU API

그룹 생성, 조회, 상세 조회, 정보 수정 API, 시리얼라이저 작성

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -75,12 +71,24 @@ def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+
+    class Meta:
+        model = CrawledLecture
+        fields = ("id", "title", "instructor", "price")
```
uuid 로 외부에 노출부탁드립니다.

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -13,14 +14,15 @@ class Meta:
             "status",
         ]
-        read_only_fields = ["id", "status"]
+        read_only_fields = ["id", "uuid", "status"]
```
여기서도 외부적으로는 uuid만 사용할거기 때문에 id는 제거해주셔도됩니다.

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -75,12 +71,24 @@ def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+    def get_price(self, obj: CrawledLecture) -> int:
+        if obj.discount_price and obj.discount_price > 0:
+            return obj.discount_price
+        return obj.original_price
```
이거 할인금액이랑 원금액이랑 따로 필드 반환하도록 해주세요

Meoyoug commented on apps/studies/views/groups.py:
```diff
@@ -49,167 +47,61 @@ def post(self, request: Request) -> Response:
+        },
+    )
+    def put(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        obj_uuid = self.kwargs.get("group_id")
```
헷갈리지않게 group_uuid로 받도록 url도함께수정바랍니다

Meoyoug commented on apps/studies/views/groups.py:
```diff
@@ -49,167 +47,61 @@ def post(self, request: Request) -> Response:
     def get(self, request: Request, *args: Any, **kwargs: Any) -> Response:
-        response_data = self.get_object()
-        return Response(response_data, status=status.HTTP_200_OK)
+        obj_uuid = self.kwargs.get("group_id")
```
헷갈리지않게 group_uuid로 받도록 url도함께수정바랍니다

Meoyoug reviewed (changes requested):
테스트코드도 빠져있네요 테스트코드 같이작성해서 PR해주세요

dev-s-kr committed:
✨ Feat: 스터디 그룹 CRU API

스터디 그룹 생성, 조회, 상세 조회, 수정 API

# Conflicts:
#	apps/studies/tests/test_groups.py

# Conflicts:
#	apps/studies/tests/test_groups.py
#	apps/studies/urls.py

dev-s-kr committed:
🎨 Style: 스터디 그룹 테스트 코드 타입 힌팅 추가

style: 스터디 그룹 테스트 코드 타입 힌팅 추가

dev-s-kr committed:
🔥 Remove: 스터디 그룹 테스트 코드 멤버 관련 코드 삭제 ( 다른 파일로 이동 )

Remove: 스터디 그룹 테스트 코드 멤버 관련 코드 삭제 ( 다른 파일로 이동 )

dev-s-kr committed:
🔥 Remove: 스터디 그룹 테스트 코드 삭제로 인해 생긴 안쓰는 임포트 삭제

remove: 스터디 그룹 테스트 코드 삭제로 인해 생긴 안쓰는 임포트 삭제

dev-s-kr committed:
🎨 Style: 스터디 그룹 API 코드 포매팅

style: 스터디 그룹 API 코드 포매팅

dev-s-kr committed:
🎨 Style: 스터디 그룹 CRU API 테스트 코드 포매팅

style: 스터디 그룹 CRU API 테스트 코드 포매팅

dev-s-kr committed:
🎨 Style: 코드 포매팅

style: 코드 포매팅

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -33,22 +34,36 @@ class StudyGroupCreateSerializer(StudyGroupBaseSerializer):
+        study_group = StudyGroup.objects.create(**validated_data)
+
+        for lecture in lectures:
+            StudyLecture.objects.create(study_group=study_group, lecture=lecture)
```
이거 반복문돌면서 create 하면 강의 수에 따라 N개의 추가쿼리가 발생하는데
한개 쿼리로 해결가능할겁니다!

Meoyoug replied:
create부분에 transaction도 묶어야하지않을까요~?

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -33,22 +34,36 @@ class StudyGroupCreateSerializer(StudyGroupBaseSerializer):
+
+        return study_group
+
+    def to_representation(self, instance: StudyGroup) -> Dict[str, Any]:
```
SlugRelatedField로 이미 uuid로 반환하게 되어있을텐데 왜 굳이 id로 다시 반환하게하셨나요?

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -33,22 +34,36 @@ class StudyGroupCreateSerializer(StudyGroupBaseSerializer):
+        ret["lectures"] = list(instance.lectures.values_list("lecture_id", flat=True))
+        return ret
+
+    def get_lectures(self, obj: StudyGroup) -> Iterable[int]:
```
lectures가 SerializerMethodField가아니라서 이건 유효하지않은 메서드인 것 같습니다

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -74,13 +89,40 @@ def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+        # 중간 테이블 처리
+        if lectures is not None:
+            StudyLecture.objects.filter(study_group=instance).delete()
+            for lecture in lectures:
```
여기서도 쿼리 반복되는 거 해결 부탁드립니다.

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -74,13 +89,40 @@ def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+        if lectures is not None:
+            StudyLecture.objects.filter(study_group=instance).delete()
+            for lecture in lectures:
+                StudyLecture.objects.create(study_group=instance, lecture=lecture)
```
이거 Lecture에 대해서 ManyToMany Relation이 적용되어있으면 set메서드로 더 간편하게 작성할 수 잇어요

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -74,13 +89,40 @@ def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+
+    class Meta:
+        model = StudyLecture
+        fields = ("uuid", "title", "instructor", "original_price", "discount_price")
```
CralwedLecture 를 Meta의 model로 두시고 Model에 ManyToManyField로 lectures 추가해주세요

Meoyoug replied:
이후에 해당 필드들 source 참조 지우셔도되고 Nested되는 시리얼라이저에서 정상적으로 동작할겁니다

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -89,46 +131,38 @@ class Meta(StudyGroupBaseSerializer.Meta):
-        ]
+    class Meta:
+        model = CrawledLecture
+        fields = ("uuid", "thumbnail_img_url", "title", "instructor", "url_link")
```
이것도 CrawledLecture에 대해서 ManyToMany 추가하시면 간결하게 작성가능할겁니다

Meoyoug commented on apps/studies/serializers/groups.py:
```diff
@@ -89,46 +131,38 @@ class Meta(StudyGroupBaseSerializer.Meta):
-    members = serializers.SerializerMethodField()
-    lectures = serializers.SerializerMethodField()
+    members = StudyGroupDetailMemberSerializer(many=True)
+    lectures = StudyGroupDetailLectureSerializer(many=True)
```
여기서 source에 추가해주시면되요 ManyToMany로 추가된 필드

Meoyoug commented on apps/studies/tests/test_groups.py:
테스트 작성시에 create를 for문으로 반복하는것 대신 bulk_create활용하기

dev-s-kr committed:
🐛 Fix: 스터디 그룹 CRU API 코드 수정

Fix: 스터디 그룹 CRU API 코드 수정

dev-s-kr committed:
🐛 Fix: 스터디 그룹 CRU API 코드 수정

fix: 스터디 그룹 CRU API 코드 수정

dev-s-kr committed:
✨ Feat: 스터디 그룹 모델 살짝 수정 및 모델에 맞게 코드 수정

Feat: 스터디 그룹 모델 살짝 수정 및 모델에 맞게 코드 수정

dev-s-kr committed:
🐛 Fix: 스터디 그룹 CRU 테스트 코드 수정

fix: 스터디 그룹 CRU 테스트 코드 수정

dev-s-kr committed:
✨ Modify model related name

- StudyLecture.lecture의 related_name을 study_lectures로 수정
- StudyLecture.study_group의 related_name을 study_lectures로 수정
- GroupMember.study_group의 related_name을 group_members로 수정
- GroupMember.user의 related_name을 group_members로 수정
- StudyGroup에 member와 lectures를 ManyToMany로 추가

dev-s-kr committed:
✅ test: 게시글 API 단위 테스트 추가

- 스터디 그룹 CRU 모델 related_name 변경( 관련 시리얼라이저, 퍼미션파일 코드도 수정 )
- url id -> uuid 로 변경
- 테스트 코드 url reverse로 변경
